### PR TITLE
Updated the github workflow to generate the JSON schemas and Definitions for Kuma components

### LINKED ISSUE #174: [CI] Patterns: Generate JSON Schema and Definitions for KUMA components
**Enhancement Description**

To generate the workflow of getting the manifest from the KUMA using the [kubeopenapi-jsonschema](https://github.com/layer5io/kubeopenapi-jsonschema) to read and generate json schemas from those manifests and then also generating workload definitions from those schemas.

A sample workflow is [here](https://github.com/DelusionalOptimist/meshery-istio/blob/gen-oam/.github/workflows/update-oam-defs.yml)

---

### PR DESCRIPTION
**Pull request by Shreyas220**
**Description**
workflow to Generate JSON Schema and Definitions for KUMA components 
This PR fixes #174

**Notes for Reviewers**

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.

### CODE REVIEW

**Comment by DelusionalOptimist** [comment]
Hey @Shreyas220! What happened here :thinking:

**Comment by Shreyas220** [comment]
nothing, i just made a mesh xD

**Comment by leecalcote** [comment]
Thanks a lot, @Shreyas220. I just soiled my computer's display.

I spit out my drink laughing.

**Comment by Shreyas220** [comment]
> Thanks a lot, @Shreyas220. I just soiled my computer's display.
> 
> I spit out my drink laughing.

damn :joy::joy::joy: 
i guess i owe you a new laptop now

File: .github/workflows/update-oam-defs.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            definitionRef=$(printf %skuma.meshery.layer5.io $nameLower)
+            name=`echo $t | tr -d '"' | tr '[:upper:]' '[:lower:]'`
+            definitionRef=$(printf %skuma.meshery.layer5.io $name)
+            ./kubeopenapi-jsonschema --location ~/kuma.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..openAPIV3Schema.properties.spec' --o-filter "$[]" -o json | jq '.[] | .version = "'$NEW_KUMA_VERSION'" | .title = "'$(printf %sKUMA $nameUpper)'"' > ../templates/oam/workloads-$NEW_KUMA_VERSION/$(printf %skuma.meshery.layer5io.schema.json $nameLower)
+            echo $template | jq ' ."metadata"."name" = "'$(printf %sKUMA $nameUpper)'" | ."spec"."definitionRef"."name"="'$definitionRef'"' > ../templates/oam/workloads-$NEW_KUMA_VERSION/$(printf %skuma_definition.json $nameLower)
+          done
+      - name: Cleanup
+        run: |
+          rm -rf util kuma-*
+      - name: Update OSM version in .env
```
Comment (sayantan1413) [inline .github/workflows/update-oam-defs.yml:77]: KUMA

**Comment by Shreyas220** [comment]
@DelusionalOptimist and @sayantan1413 could you check i added the separators 
https://github.com/Shreyas220/meshery-kuma/tree/master/templates/oam/workloads-1.2.2

**Comment by DelusionalOptimist** [comment]
@Shreyas220 this may feel like nitpicking but `KUMA` is `Kuma` : )

**Comment by DelusionalOptimist** [comment]
Seems like some of the extracted schemas have lost. Will you be able to investigate this? :thinking: 
For e.g: https://github.com/Shreyas220/meshery-kuma/blob/master/templates/oam/workloads-1.2.2/dataplaneinsight.kuma.meshery.layer5io.schema.json

**Comment by Shreyas220** [comment]
> ataplaneinsight.kuma.meshery.layer5io.schema.json

looks like dataplaneinsight.kuma.meshery.layer5io.schema.json is the only one lost :thinking::thinking::thinking:

**Comment by Shreyas220** [comment]
@DelusionalOptimist https://github.com/kumahq/kuma/blob/master/deployments/charts/kuma/crds/dataplane-insight.yaml this the yaml in kuma repo

i think in the fourth last line `status ` should be `spec `
please confirm

**Comment by DelusionalOptimist** [comment]
Yeah good catch @Shreyas220. That might be casuing it. I think we need to take a look at the structure of kuma crds and modify the jsonpath expression accordingly.

**Comment by DelusionalOptimist** [comment]
This might or might not fix `dataplaneinsight.kuma.meshery.layer5io.schema.json` but seems like we aren't extracting the necessary info from the manifests.
For e.g.: we'll need all this stuff
https://github.com/kumahq/kuma/blob/4d7e7c03bab0c72fd2cbb77d82df97929dfd096c/deployments/charts/kuma/crds/circuit-breaker.yaml#L18-L25

**Comment by Shreyas220** [comment]
> Yeah good catch @Shreyas220. That might be casuing it. I think we need to take a look at the structure of kuma crds and modify the jsonpath expression accordingly.

yes or.......................a simpler approach(by which I mean easier (not because I am lazy ...nope ...never)) can be  we can change `status` to `spec`  in the kuma repo ?

**Comment by Shreyas220** [comment]
> Yeah good catch @Shreyas220. That might be casuing it. I think we need to take a look at the structure of kuma crds and modify the jsonpath expression accordingly.

comparing with other yaml file, `dataplaneinsight` is the only one where `status` is used

File: .github/workflows/update-oam-defs.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+          cd util
+          template='{"apiVersion":"core.oam.dev/v1alpha1","kind":"WorkloadDefinition","metadata":{"name":""},"spec":{"definitionRef":{"name":""}}}'
+          crds=$(./kubeopenapi-jsonschema --location ~/kuma.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition")]' -o json --o-filter '$..["spec"]["names"]["kind"]' | jq '.[]')
+          for t in ${crds[@]}; do
+            nameUpper=`echo $t  | tr -d '"'`
+            nameLower=`echo $t  | tr -d '"' | tr '[:upper:]' '[:lower:]'`
+            definitionRef=$(printf %s.kuma.meshery.layer5.io $nameLower)
+            name=`echo $t | tr -d '"' | tr '[:upper:]' '[:lower:]'`
+            definitionRef=$(printf %s.kuma.meshery.layer5.io $name)
+            ./kubeopenapi-jsonschema --location ~/kuma.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..openAPIV3Schema.properties.spec' --o-filter "$[]" -o json | jq '.[] | .version = "'$NEW_KUMA_VERSION'" | .title = "'$(printf %sKUMA $nameUpper)'"' > ../templates/oam/workloads-$NEW_KUMA_VERSION/$(printf %s.kuma.meshery.layer5io.schema.json $nameLower)
```
Comment (DelusionalOptimist) [inline .github/workflows/update-oam-defs.yml:71]:
```suggestion
            ./kubeopenapi-jsonschema --location ~/kuma.yaml -t yaml --filter '$[?(@.kind=="CustomResourceDefinition" && @.spec.names.kind=='$t')]..openAPIV3Schema' --o-filter "$[]" -o json | jq '.[] | .version = "'$NEW_KUMA_VERSION'" | .title = "'$(printf %sKUMA $nameUpper)'"' > ../templates/oam/workloads-$NEW_KUMA_VERSION/$(printf %s.kuma.meshery.layer5io.schema.json $nameLower)
```
Try using this. This should fix it.

**Comment by Shreyas220** [comment]
working

File: .github/.env [hunk 3]
```diff
@@ -0,0 +1 @@
+CURRENT_KUMA_VERSION= 1.2.2
```
Comment (leecalcote) [inline .github/.env:1]: Version numbers need to be embedded in the components themselves. Let’s avoid tracking version numbers separately. // @utkarsh-pro @DelusionalOptimist

**Review by leecalcote — CHANGES_REQUESTED** [review]

File: .github/.env (hunk 3 above)
Comment (DelusionalOptimist) [reply .github/.env:1]: Cool :+1:

**Comment by leecalcote** [comment]
Have we documented our revised approach to versioning, @DelusionalOptimist? This [design spec](https://docs.google.com/document/d/1b8JAMzr3Rntu7CudRaYv6r6ccACJONAB5t7ISCaPNuA/edit) or the patterns design spec could be used.

**Comment by DelusionalOptimist** [comment]
> Have we documented our revised approach to versioning, @DelusionalOptimist? This design spec or the patterns design spec could be used.

@leecalcote You mean the one about dealing with different API versions of the same resource? Nope. We're just targeting the 'lexicographically' first resource for now. I'm not sure if the current design implements support for all of them. Will need @utkarsh-pro's thoughts.
Also see: https://layer5io.slack.com/archives/CFGG6U10E/p1627907677034300?thread_ts=1627747909.020600&cid=CFGG6U10E

**Comment by DelusionalOptimist** [comment]
@sayantan1413 will you be able to guide @Shreyas220 in carrying this forward? :slightly_smiling_face:

**Comment by sayantan1413** [comment]
Yeah sure @DelusionalOptimist I will guide @Shreyas220 .

**Pull request closed without merging**